internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubProducerAdapterConcurrentDelegator.java

```diff
@@ -0,0 +1,279 @@
+    if (producerThreadCount <= 1) {
+      throw new VeniceException("'producerThreadCount' should be larger than 1, but got " + producerThreadCount);
+    }
+    if (producerQueueSize <= 2 * 1024 * 1024) {
+      throw new VeniceException("'producerQueueSize' should be larger than 2MB, but got " + producerQueueSize);
+    }
+    LOGGER.info(
+        "Initializing a 'PubSubProducerAdapterConcurrentDelegator' instance for topic: {} with thread"
+            + " count: {} and queue size: {}",
+        producingTopic,
+        producerThreadCount,
+        producerQueueSize);
+    this.blockingQueueList = new ArrayList<>(producerThreadCount);
+    this.producerExecutor = Executors
+        .newFixedThreadPool(producerThreadCount, new DaemonThreadFactory(producingTopic + "-concurrent-producer"));
+    this.partitionQueueAssignment = new VeniceConcurrentHashMap<>();
+    this.lastNodePerQueue = new VeniceConcurrentHashMap<>(producerThreadCount);
+    this.producingTopic = producingTopic;
+    this.producers = new ArrayList<>(producerThreadCount);
+    this.drainers = new ArrayList<>(producerThreadCount);
+    for (int cur = 0; cur < producerThreadCount; ++cur) {
+      PubSubProducerAdapter producer = producerAdapterSupplier.get();
+      this.producers.add(producer);
+
+      MemoryBoundBlockingQueue<ProducerQueueNode> blockingQueue =
+          new MemoryBoundBlockingQueue<>(producerQueueSize, 1 * 1024 * 1024l);
+      ProducerQueueDrainer drainer = new ProducerQueueDrainer(blockingQueue, cur, producer);
+      this.blockingQueueList.add(blockingQueue);
+      this.drainers.add(drainer);
+      this.producerExecutor.submit(drainer);
+    }
+  }
+
+  public static class ProducerQueueNode implements Measurable {
+    /**
+     * Considering the overhead of {@link PubSubMessage} and its internal structures.
+     */
+    private static final int QUEUE_NODE_OVERHEAD_IN_BYTE = 256;
+    private final String topic;
+    private final int partition;
+    private final KafkaKey key;
+    private final KafkaMessageEnvelope value;
+    private final PubSubMessageHeaders headers;
+    private final PubSubProducerCallback callback;
+    private final CompletableFuture<PubSubProduceResult> produceFuture;
+
+    public ProducerQueueNode(
+        String topic,
+        int partition,
+        KafkaKey key,
+        KafkaMessageEnvelope value,
+        PubSubMessageHeaders headers,
+        PubSubProducerCallback callback,
+        CompletableFuture<PubSubProduceResult> produceFuture) {
+      this.topic = topic;
+      this.partition = partition;
+      this.key = key;
+      this.value = value;
+      this.headers = headers;
+      this.callback = callback;
+      this.produceFuture = produceFuture;
+    }
+
+    @Override
+    public int getSize() {
+      int headerSize = 0;
+      if (headers != null) {
+        for (PubSubMessageHeader header: headers.toList()) {
+          headerSize += header.key().length() + header.value().length;
```

Comment:
Added a new dependency to measure the on-heap object size:
https://github.com/openjdk/jol

Reply:
I had to revert the jol dependency as it would crash mac test run and stuck CI run.
One related bug report:
https://bugs.openjdk.org/browse/CODETOOLS-7903447

Reply:
That's fine, no need to integrate that right now. Thanks!